lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -0,0 +1,174 @@
+"""
+pbench-clear-tools
+
+This script will remove tools that have been registered.  If no options are
+used, then all tools from the "default" tool group are removed.  Specifying
+a tool name and/or remote host will limit the scope of the removal.
+
+
+"""
+import shutil
+import sys
+import os
+import pathlib
+
+import click
+
+from pbench.agent.utils import setup_logging
+from pbench.cli.agent import CliContext, pass_cli_context
+from pbench.cli.agent.commands.tools.base import ToolCommand
+from pbench.cli.agent.options import common_options
+
+
+class ClearTools(ToolCommand):
+    """Clear registered tools"""
+
+    def __init__(self, context):
+        super().__init__(context)
+
+        self.logger = setup_logging(logfile=self.pbench_log)
+
+    def execute(self):
+        errors = 0
+
+        self.tool_group_dir = self.tool_group_dir(self.context.group)
+        if not self.tool_group_dir.exists():
+            return 1
+
+        try:
+            self.context.remote = self.context.remote.split(",")
+        except Exception:
+            # We were not given any remotes on the command line, build the list
+            # from the tools group directory.
+            self.context.remote = self.remote(self.tool_group_dir)
+
+        for remote in self.context.remote:
+            tg_dir_r = self.tool_group_dir / remote
+            if not tg_dir_r.exists():
+                self.logger.warn(
+                    f'The given remote host, "{remote}", is not a directory in {self.tool_group_dir}.'
+                )
+                continue
+
+            if not self.context.name:
+                # Discovery all the tools registered for this remote
+                self.context.name = self.tools(tg_dir_r)
+
+            if self.context.name:
+                for name in self.context.name:
+                    status = self._clear_tools(name, remote)
+                    if status != 0:
+                        errors += 1
+
+            tool_files = [p.name for p in tg_dir_r.iterdir()]
+
+            if "__label__" in tool_files:
+                label = tg_dir_r / "__label__"
+                try:
+                    os.unlink(label)
+                except OSError:
+                    self.logger.error(f"Failed to remove label for remote {tg_dir_r}")
+                    errors += 1
+
+            if self.is_empty(tg_dir_r):
+                self.logger.info(f'All tools removed from host, "{tg_dir_r.name}"')
+                try:
+                    shutil.rmtree(tg_dir_r)
+                except OSError:
+                    self.logger.error(f"Failed to remove remote directory {tg_dir_r}")
+                    errors += 1
+
+        return errors
+
+    def _clear_tools(self, name, remote):
+        """Remove specified tool and associated files"""
+        tpath = self.tool_group_dir / remote / name
+        try:
```

Comment:
Perhaps we could consider a more pythonic method where we don't perform the tests, we try the removal first?

```
try:
    tpath.unlink()
except Exception as exc:
    self.logger.error("Failed to remove %s: %s", tpath, exc)
    ret_val = 1
else:
    noinstall = pathlib.Path(f"{tpath}.__noinstall__")
    try:
        noinstall.unlink()
    except FileNotFoundError:
        ret_val = 0
    except Exception as exc:
        self.logger.error("Failure to remove %s: %s", noinstall, exc)
        ret_val = 1
    else:
        ret_val = 0

if ret_val == 0:
    self.logger.info('Remove "%s" from host, "%s", in tools group, "%s"', name, remote, self.context.group)
return ret_val
```

And we also want to allow the logging subsystem to format the log messages when appropriate.

Reply:
done

Reply:
fixed